src/graphapi/enginefinder.cpp

```diff
@@ -0,0 +1,167 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/errors.hpp>
+#include <miopen/graphapi/enginefinder.hpp>
+
+#include <algorithm>
+
+void miopen::graphapi::BackendOperationGraphDescriptor::setAttribute(
+    miopenBackendAttributeName_t attributeName,
+    miopenBackendAttributeType_t attributeType,
+    int64_t elementCount,
+    void* arrayOfElements)
+{
+    if(mFinalized)
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
+
+    switch(attributeName)
+    {
+    case MIOPEN_ATTR_OPERATIONGRAPH_HANDLE:
+        if(attributeType == MIOPEN_TYPE_HANDLE && elementCount == 1)
+        {
+            mHandle = *static_cast<miopenHandle_t*>(arrayOfElements);
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+        break;
+
+    case MIOPEN_ATTR_OPERATIONGRAPH_OPS:
+        if(attributeType == MIOPEN_TYPE_BACKEND_DESCRIPTOR && elementCount > 0)
+        {
+            std::vector<miopenBackendDescriptor_t> ops;
+            ops.reserve(elementCount);
+
+            OpGraphBuilder builder;
+
+            std::for_each_n(static_cast<miopenBackendDescriptor_t*>(arrayOfElements),
+                            elementCount,
+                            [&ops, &builder](miopenBackendDescriptor_t apiDescriptor) {
+                                BackendDescriptor& backendDescriptor = deref(apiDescriptor);
+                                if(backendDescriptor.isFinalized())
+                                {
+                                    OpNode* node = backendDescriptor.getOperation();
+                                    if(node != nullptr && !builder.hasNode(node))
+                                    {
+                                        builder.addNode(node);
+                                        ops.push_back(apiDescriptor);
+                                        return;
+                                    }
+                                }
+                                MIOPEN_THROW(miopenStatusBadParm);
+                            });
+
+            mOps            = std::move(ops);
+            mOpGraphBuilder = std::move(builder);
```

Comment:
This does not seem right. A user can add multiple nodes by calling `setAttribute` multiple times and we should keep accumulating them.

Reply:
Disagree. The attribute is set in one step by passing the whole array of nodes.

Reply:
Not sure if the API documentation says that. There can be situations where we cannot add all nodes at once or don't want to add all nodes at once.

Reply:
But then maybe you're right. `Set` semantics would mean that you do it in one go. That's just bad API design but let's go with this.